File: packages/opencode/src/bastion/audit.ts
import fs from "fs"
import path from "path"
import crypto from "crypto"
import { Instance } from "@/project/instance"
import { Log } from "@/util/log"

const log = Log.create({ service: "bastion-audit" })

export interface AuditEntry {
  timestamp: string
  session_id: string
  event_id: string
  tool_name: string
  tool_input: Record<string, any>
  status: "SAFE" | "UNSAFE" | "BLOCKED" | "EXECUTED" | "SANITIZED_EXECUTED" | "LEARNED"
  enforcement: string
  risk_reason: string
  rule_matched: string | null
  sanitized_command: Record<string, any> | null
  execution_result: string | null
  retry_number: number
  user_request: string
}

const SENSITIVE_KEYS = ["password", "token", "key", "secret", "api_key"]

function auditPath(): string {
  return path.join(Instance.directory, ".opencode", "bastion_audit.json")
}

function redact(toolInput: Record<string, any>): Record<string, any> {
  const result: Record<string, any> = {}
  for (const [k, v] of Object.entries(toolInput)) {
    if (SENSITIVE_KEYS.some((s) => k.toLowerCase().includes(s))) {
      result[k] = "[REDACTED]"
    } else {
      result[k] = v
    }
  }
  return result
}

export namespace BastionAudit {
  export function record(input: {
    sessionID: string
    toolName: string
    toolInput: Record<string, any>
    status: AuditEntry["status"]
    enforcement: string
    riskReason: string
    ruleMatched?: string | null
    sanitizedCommand?: Record<string, any> | null
    executionResult?: string | null
  }): void {
    try {
      const p = auditPath()
      const dir = path.dirname(p)
      if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true })

      let entries: AuditEntry[] = []
      try {
        if (fs.existsSync(p)) {
          entries = JSON.parse(fs.readFileSync(p, "utf-8"))
        }
      } catch {
        entries = []
      }

      const entry: AuditEntry = {
        timestamp: new Date().toISOString(),
        session_id: input.sessionID,
        event_id: `evt_${crypto.randomBytes(4).toString("hex")}`,
        tool_name: input.toolName,
        tool_input: redact(input.toolInput),
        status: input.status,
        enforcement: input.enforcement,
        risk_reason: input.riskReason,
        rule_matched: input.ruleMatched ?? null,
        sanitized_command: input.sanitizedCommand ?? null,
        execution_result: input.executionResult ? input.executionResult.slice(0, 500) : null,
        retry_number: 0,
        user_request: "",
      }

      entries.push(entry)
      fs.writeFileSync(p, JSON.stringify(entries, null, 2))
    } catch (e) {
      log.error("failed to write bastion audit", { error: e })
    }
  }

  export function getAll(): AuditEntry[] {
    try {
      const p = auditPath()
      if (!fs.existsSync(p)) return []
      return JSON.parse(fs.readFileSync(p, "utf-8"))
    } catch {
      return []
    }
  }

  export function clear(): void {
    try {
      const p = auditPath()
      const dir = path.dirname(p)
      if (!fs.existsSync(dir)) fs.mkdirSync(dir, { recursive: true })
      fs.writeFileSync(p, "[]")
    } catch (e) {
      log.error("failed to clear bastion audit", { error: e })
    }
  }
}

File: sentinel-sdk/audit.py
"""
Sentinel-SDK: Audit Logger
Append-only JSON log for all tool call events.
"""

import json
import os
import uuid
from datetime import datetime, timezone


class AuditLogger:
    def __init__(self, audit_path: str):
        self.audit_path = audit_path
        if not os.path.exists(self.audit_path):
            with open(self.audit_path, 'w') as f:
                json.dump([], f)

    def log(self, session_id: str, tool_name: str, tool_input: dict,
            status: str, enforcement: str, risk_reason: str,
            rule_matched: str = None, user_request: str = "",
            retry_number: int = 0, sanitized_command: dict = None,
            execution_result: str = None):
        entry = {
            "timestamp": datetime.now(timezone.utc).isoformat(),
            "session_id": session_id,
            "event_id": f"evt_{uuid.uuid4().hex[:8]}",
            "tool_name": tool_name,
            "tool_input": self._redact(tool_input),
            "status": status,
            "enforcement": enforcement,
            "risk_reason": risk_reason,
            "rule_matched": rule_matched,
            "sanitized_command": sanitized_command,
            "execution_result": (execution_result[:500]
                                 if execution_result else None),
            "retry_number": retry_number,
            "user_request": user_request
        }

        try:
            with open(self.audit_path) as f:
                logs = json.load(f)
        except (json.JSONDecodeError, FileNotFoundError):
            logs = []

        logs.append(entry)

        with open(self.audit_path, 'w') as f:
            json.dump(logs, f, indent=2)

    def _redact(self, tool_input: dict) -> dict:
        redacted = {}
        sensitive_keys = ["password", "token", "key", "secret", "api_key"]
        for k, v in tool_input.items():
            if any(s in k.lower() for s in sensitive_keys):
                redacted[k] = "[REDACTED]"
            else:
                redacted[k] = v
        return redacted

    def get_all(self) -> list:
        try:
            with open(self.audit_path) as f:
                return json.load(f)
        except (json.JSONDecodeError, FileNotFoundError):
            return []

    def clear(self):
        with open(self.audit_path, 'w') as f:
            json.dump([], f)
